File: Source/lua-common/LuaThread.cpp
//
// Created by 冯鸿杰 on 2018/12/25.
//

#include "LuaThread.h"
#include "LuaEngineAdapter.hpp"
#include "LuaSession.h"
#include "StringUtils.h"
#include "LuaDataExchanger.h"
#include "LuaTuple.h"
#include "LuaExportsTypeManager.hpp"

using namespace cn::vimfung::luascriptcore;

LuaThread::LuaThread(LuaContext *context, LuaMethodHandler handler)
    : LuaContext(context -> getExportsTypeManager() -> getPlatform(), [this, context] () -> lua_State* {
    return LuaEngineAdapter::newThread(context -> getCurrentSession() -> getState());
})
{
    _context = context;
    _finished = false;

    _exchangeId = StringUtils::format("%p", this);

    int top = LuaEngineAdapter::getTop(_context -> getCurrentSession() -> getState());
    _context -> getDataExchanger() -> setLuaObject(top, _exchangeId);
    _context -> getDataExchanger() -> retainLuaObject(this);
    LuaEngineAdapter::pop(_context -> getCurrentSession() -> getState(), 1);

    registerMethod("_handler", handler);
}

LuaThread::~LuaThread()
{
    _context -> getDataExchanger() -> releaseLuaObject(this);
}

LuaContext* LuaThread::getContext()
{
    return _context;
}

bool LuaThread::hasFinished()
{
    return _finished;
}

void LuaThread::resume(LuaArgumentList argumentList)
{
    getGlobal("_handler");

    for (LuaArgumentList::iterator it = argumentList.begin(); it != argumentList.end() ; it++)
    {
        LuaValue *arg = *it;
        arg -> push(this);
    }

    int status = LuaEngineAdapter::resumeThread(getCurrentSession() -> getState(), NULL, (int)argumentList.size());
    _finished = status != LUA_YIELD;
}

void LuaThread::yield(LuaValue *resultValue)
{
    int resultCount = 1;
    if (resultValue -> getType() == LuaValueTypeTuple)
    {
        resultCount = (int)resultValue -> toTuple() -> count();
    }

    resultValue -> push(this);
    LuaEngineAdapter::yieldThread(getCurrentSession() -> getState(), resultCount);
}

File: Source/Android/luascriptcore/src/main/java/cn/vimfung/luascriptcore/LuaThread.java
package cn.vimfung.luascriptcore;

public class LuaThread extends LuaBaseObject
{
    private LuaContext _context;
    private LuaMethodHandler _handler;
    private boolean _finished;

    /**
     * 初始化
     * @param nativeId 原生对象标志
     */
    protected LuaThread(int nativeId)
    {
        super(nativeId);
    }

    /**
     * 创建线程
     * @param context 上下文对象
     * @param handler 线程处理器
     * @return 线程对象
     */
    public static LuaThread create(LuaContext context, LuaMethodHandler handler)
    {
        LuaThread thread = LuaNativeUtil.createThread(context);
        thread._context = context;
        thread._handler = handler;

        return thread;
    }

    /**
     * 恢复线程
     * @param arguments 请求参数列表
     */
    public void resume(LuaValue[] arguments)
    {
        LuaNativeUtil.resumeThread(this, arguments);
    }

    /**
     * 挂起线程
     * @param resultValue 返回值
     */
    public void yield(LuaValue resultValue)
    {
        LuaNativeUtil.yieldThread(this, resultValue);
    }

    /**
     * 获取上下文对象
     * @return 上下文对象
     */
    public LuaContext getContext()
    {
        return _context;
    }

    /**
     * 是否完成
     * @return true 表示执行完成，false 表示尚未完成
     */
    public boolean hasFinished()
    {
        return _finished;
    }

    /**
     * 调用方法
     * @param methodName    方法名称
     * @param arguments     方法的传入参数
     * @return              返回值
     */
    private LuaValue methodInvoke (String methodName, LuaValue[] arguments)
    {
        _handler.onExecute(arguments);
        return null;

    }
}
